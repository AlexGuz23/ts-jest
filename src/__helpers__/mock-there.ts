export default function mockThese(map: string[] | { [k: string]: () => any }) {
  const isArray = Array.isArray(map);
  const items: string[] = isArray ? (map as string[]) : Object.keys(map);
  items.forEach(item => {
    const val = isArray ? () => item : map[item];
    jest.doMock(item, val, { virtual: true });
  });
}
