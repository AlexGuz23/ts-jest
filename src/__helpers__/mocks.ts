import { rootLogger } from '../util/logger'
import { testing } from 'bs-logger'

// typings helper
export function mocked<T>(
  val: T,
): T extends (...args: any[]) => any ? jest.MockInstance<T> : jest.Mocked<T> {
  return val as any
}
export function spied<T>(
  val: T,
): T extends (...args: any[]) => any ? jest.SpyInstance<T> : jest.Mocked<T> {
  return val as any
}

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target
