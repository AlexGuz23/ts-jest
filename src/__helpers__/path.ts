import { resolve, relative } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function relativeToRoot(path: string) {
  return relative(ROOT, path)
}
