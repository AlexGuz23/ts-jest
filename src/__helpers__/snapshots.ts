import outdent from 'outdent'

export const snap = outdent({
  trimTrailingNewline: false,
  trimLeadingNewline: false,
})
