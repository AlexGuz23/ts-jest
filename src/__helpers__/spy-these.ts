export default function spyThese<T extends object, K extends keyof T>(
  object: T,
  implementations: { [key in K]: T[K] | any | undefined },
): { [key in K]: jest.SpyInstance<T[K]> } & { mockRestore: () => void } {
  const keys = Object.keys(implementations) as K[];
  const res = keys.reduce(
    (map, key) => {
      const actual = object[key] as any;
      const spy = jest.spyOn(object, key as K);
      if (implementations[key]) {
        const impl = implementations[key] as (...args: any[]) => any;
        if (impl.length && /\W\$super\W/.test(impl.toString())) {
          spy.mockImplementation(function(this: T, ...args: any[]) {
            return impl.call(this, () => actual.apply(this, args), ...args);
          });
        } else {
          spy.mockImplementation(impl);
        }
      }
      return map;
    },
    {} as any,
  );
  // utility to restore all
  res.mockRestore = () => {
    keys.forEach(key => res[key].mockRestore());
  };
  return res;
}
