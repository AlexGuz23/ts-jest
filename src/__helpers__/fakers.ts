import { TsJestGlobalOptions, BabelConfig } from '../types';
import { resolve } from 'path';
import { ImportReasons } from '../utils/messages';

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath);
}

export const rootDir = filePath('');

export function transpiledTsSource() {
  return `
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var upper_1 = __importDefault(require("./upper"));
var lower_1 = __importDefault(require("./lower"));
jest.mock('./upper', function () { return function (s) { return s.toUpperCase(); }; });
describe('hello', function () {
    test('my test', function () {
        expect(upper_1.default('hello')).toBe('HELLO');
        expect(lower_1.default('HELLO')).toBe('hello');
        jest.mock('./lower', function () { return function (s) { return s.toLowerCase(); }; });
    });
});
`;
}

export function htmlSource() {
  return `
<div>
  <span>some text with \`backtick\`</span>
</div>
`;
}

export function typescriptSource() {
  return `
import upper from './upper';
import lower from './lower';

jest.mock('./upper', () => (s) => s.toUpperCase());

describe('hello', () => {
  test('my test', () => {
    expect(upper('hello')).toBe('HELLO');
    expect(lower('HELLO')).toBe('hello');
    jest.mock('./lower', () => (s) => s.toLowerCase());
  });
});
`;
}

export function tsJestConfig<T extends TsJestGlobalOptions>(
  options?: TsJestGlobalOptions,
): T {
  return { ...options } as any;
}

export function jestConfig<T extends jest.ProjectConfig>(
  options?: jest.InitialOptions,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    moduleFileExtensions: ['ts', 'js'],
    ...options,
  } as any;
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestConfig(tsJestOptions);
  }
  return res;
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any;
}

export function importReason(text: string = 'because'): ImportReasons {
  return text as any;
}
