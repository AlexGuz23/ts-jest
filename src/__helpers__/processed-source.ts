import { RawSourceMap } from 'source-map'
import {
  rewriteSourceMaps,
  relativisePaths,
  parseSource,
  ParsedSourceWithMaps,
} from './source-maps'
import { ROOT } from './path'
import { isAbsolute, relative } from 'path'

// tslint:disable-next-line:no-default-export
export default class ProcessedSource {
  readonly filename: string

  constructor(
    readonly output: string,
    filename: string,
    readonly cwd: string = ROOT,
  ) {
    if (isAbsolute(filename)) filename = relative(this.cwd, filename)
    this.filename = filename
  }
  get outputCode(): string {
    return this.output
  }
  get parsedSource(): ParsedSourceWithMaps {
    return parseSource(this.output)
  }
  get outputCodeWithoutMaps(): string {
    return this.parsedSource.source
  }
  get outputSourceMaps(): RawSourceMap | undefined {
    return this.parsedSource.sourceMaps
  }
  get normalizedOutputCode(): string {
    return rewriteSourceMaps(this.output, this.sourceMapsNormalizer)
  }
  get normalizedOutputSourceMaps(): RawSourceMap | undefined {
    const maps = this.outputSourceMaps
    if (maps) return this.sourceMapsNormalizer(maps)
    return
  }
  get sourceMapsNormalizer() {
    return (maps: RawSourceMap): RawSourceMap =>
      relativisePaths(maps, this.cwd, '<cwd>/')
  }
}
