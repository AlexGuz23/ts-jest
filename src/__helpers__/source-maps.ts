// WARNING: this file is shared between e2e and unit tests
import { RawSourceMap } from 'source-map'
import bufferFrom from 'buffer-from'
import { relative, isAbsolute, resolve } from 'path'
import stableStringify = require('fast-json-stable-stringify')
import { realpathSync } from 'fs'

// source-map module doesn't provide a sync method to extract the source-maps
export function extractSourceMaps(source: string): RawSourceMap | undefined {
  const [, comment]: [any, string | undefined] =
    (source.match(
      /[\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,(\S+)\s*$/,
    ) as any) || []
  if (!comment) return

  return base64ToSourceMaps(comment)
}

export function base64ToSourceMaps(base64: string): RawSourceMap {
  return JSON.parse(bufferFrom(base64, 'base64').toString('utf8'))
}

export function sourceMapsToBase64(sourceMaps: RawSourceMap): string {
  return bufferFrom(stableStringify(sourceMaps)).toString('base64')
}

export function relativiseSourceRoot(
  fromPath: string,
  source: string,
  prefix: string = '',
): string {
  const from = realpathSync(fromPath)
  const remap = (path: string): string =>
    (isAbsolute(path)
      ? `${prefix}${relative(from, realpathSync(path))}`
      : path
    ).replace(/\\/g, '/')

  return source.replace(
    /([\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,)(\S+)(\s*)$/,
    (_, before, base64, after) => {
      const map = base64ToSourceMaps(base64)
      if (map.sourceRoot) map.sourceRoot = remap(map.sourceRoot)
      if (map.sources) map.sources = map.sources.map(remap)
      if (map.file) map.file = remap(map.file)
      return `${before}${sourceMapsToBase64(map)}${after}`
    },
  )
}
