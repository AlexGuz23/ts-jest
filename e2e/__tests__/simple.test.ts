import { configureTestCase } from '../__helpers__/test-case'
import { allPackageSets } from '../__helpers__/templates'

describe('Simple test', () => {
  const testCase = configureTestCase('simple')

  testCase.runWithTemplates(allPackageSets, 0, (runTest, { testLabel }) => {
    it(testLabel, () => {
      const result = runTest()
      expect(result.status).toBe(0)
      expect(result).toMatchSnapshot()
    })
  })
})
