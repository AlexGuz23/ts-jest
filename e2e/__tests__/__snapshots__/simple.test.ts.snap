// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple test should pass using template "default" 1`] = `
  √ jest
  ↳ exit code: 0
  ===[ STDOUT ]===================================================================
  
  ===[ STDERR ]===================================================================
  PASS ./Hello.spec.ts
    Hello Class
      √ should create a new Hello
  
  Test Suites: 1 passed, 1 total
  Tests:       1 passed, 1 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`Simple test should pass using template "with-babel-6" 1`] = `
  √ jest
  ↳ exit code: 0
  ===[ STDOUT ]===================================================================
  
  ===[ STDERR ]===================================================================
  PASS ./Hello.spec.ts
    Hello Class
      √ should create a new Hello
  
  Test Suites: 1 passed, 1 total
  Tests:       1 passed, 1 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`Simple test should pass using template "with-babel-7" 1`] = `
  √ jest
  ↳ exit code: 0
  ===[ STDOUT ]===================================================================
  
  ===[ STDERR ]===================================================================
  PASS ./Hello.spec.ts
    Hello Class
      √ should create a new Hello
  
  Test Suites: 1 passed, 1 total
  Tests:       1 passed, 1 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`Simple test should pass using template "with-jest-22" 1`] = `
  √ jest
  ↳ exit code: 0
  ===[ STDOUT ]===================================================================
  
  ===[ STDERR ]===================================================================
  PASS ./Hello.spec.ts
    Hello Class
      √ should create a new Hello
  
  Test Suites: 1 passed, 1 total
  Tests:       1 passed, 1 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`Simple test should pass using template "with-typescript-2-7" 1`] = `
  √ jest
  ↳ exit code: 0
  ===[ STDOUT ]===================================================================
  
  ===[ STDERR ]===================================================================
  PASS ./Hello.spec.ts
    Hello Class
      √ should create a new Hello
  
  Test Suites: 1 passed, 1 total
  Tests:       1 passed, 1 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;
